### PR TITLE
feat(list) qri.repository as a dict with tab-completion, html repr

### PR DESCRIPTION
**Pull request by orblivion**
It sounded like you were looking for a better HTML representation for, and possibly a rewrite of, qri.list(). I had these ideas, wondering what you think.

Instead of having qri.list(), I just put a subclassed dict at `qri.repository`. That way, they can just type `ds = qri.repository[` and then tab-complete for the ref names. We could just as well do `repo = qri.repository(); ds = repo[` and tab-complete, but it seems unnecessarily verbose. If we could tab-complete `qri.get(` we'd just do that but iPython/Jupyter don't seem so extensible.

I also made an attempt at a html repr.

qri repository

Problem is that if you do `qri.repository['` (including the opening quote) before hitting tab, you get a bunch of junk. The usefulness requires specific habits on the part of the user. I was originally hoping for something like:

`qri.repository.b5.world_bank_population`

which is great for tab completion, but then I realized you allow dashes in the names.

### CODE REVIEW

**Comment by orblivion** [comment]
Threw in title. I was wondering what would happen if the title was super long so I did that as a test. Seems okay, also attached.
table2
table3

**Comment by dustmop** [comment]
The issues around tab-completion are unfortunate; the fact that users need to type `'` makes it less ergonomic and less discoverable, even when using this alternative approach to listing.

Importantly, we don't want to get rid of the `qri.list()` command, as we want to match the command name already used by the qri command-line client, and the http api. Having two different names to do nearly the same thing has caused problems in the past.

However, the html display for a list of datasets looks pretty swell. I think we should try cutting off a title if it runs over a certain length, replacing the ending with ellipsis (...).

**Comment by orblivion** [comment]
Actually they just have to type `qri.repository[`, they don't need the `'`.
My point was that if they _do_ type the`'` and hit tab, they get a bunch
more junk before the useful stuff.

**Comment by orblivion** [comment]
Actually they just have to type `qri.repository[`, they don't need the `'`. My point was that if they _do_ type the`'` and hit tab, they get a bunch more junk before the useful stuff.

(sorry I had to re-send this from Github's interface because it refused to accept Markdown from email)

**Comment by dustmop** [comment]
Yes, that's what I was referring to with the mention of ergonomics, it's some hidden gotchas that make it slightly harder to use if you don't know about it. For the reason that this is so similar to `list`, but slightly different in how it works, I don't think this is something we want to add just now.

**Comment by orblivion** [comment]
qri-list

File: qri/dataset.py [hunk 1]
```diff
@@ -134,5 +136,28 @@ def __repr__(self):
...
-        content = ', '.join(['%s' % d for d in self])
-        return '<code>[%s]</code>' % (content,)
+        dss_by_user = defaultdict(list)
+        for ds in self:
+            dss_by_user[ds.username].append(ds)
+
+        rows = ""
+        for idx, (username, dss) in enumerate(dss_by_user.items()):
+            for ds in dss:
+                meta_title = ds.versionInfo.meta_title or '(none)'
```
Comment (orblivion) [inline qri/dataset.py:146]: Depending on how attached you are to this, we now have a reason to revisit the `versionInfo` question.

File: qri/dataset.py [hunk 2]
```diff
@@ -131,6 +135,13 @@ def meta(self):
         self._ensure_populated()
         return self.meta_component
 
+    @property
+    def title(self):
```
Comment (orblivion) [inline qri/dataset.py:139]: So I avoiding this:

https://github.com/qri-io/qri-python/pull/37#discussion_r497213156

by doing this awful thing instead. Not sure which you deem worse. Or if you have a better idea.

Comment (orblivion) [reply qri/dataset.py:139]: I suppose I should write a test for this if it's a keeper.

**Comment by dustmop** [comment]
I think this is really good, as is. I'll merge and can follow up with tests and some cleanup. Thank you for your hard work!

**Pull request merged**